src/diffusers/models/controlnetxs.py

```diff
@@ -0,0 +1,1009 @@
+            encoder_hid_dim_type=encoder_hid_dim_type,
+            attention_head_dim=attention_head_dim,
+            num_attention_heads=num_attention_heads,
+            dual_cross_attention=dual_cross_attention,
+            use_linear_projection=use_linear_projection,
+            upcast_attention=upcast_attention,
+            resnet_time_scale_shift=resnet_time_scale_shift,
+            resnet_skip_time_act=resnet_skip_time_act,
+            resnet_out_scale_factor=resnet_out_scale_factor,
+            time_embedding_type=time_embedding_type,
+            time_embedding_dim=time_embedding_dim,
+            time_embedding_act_fn=time_embedding_act_fn,
+            timestep_post_act=timestep_post_act,
+            time_cond_proj_dim=time_cond_proj_dim,
+            conv_in_kernel=conv_in_kernel,
+            conv_out_kernel=conv_out_kernel,
+            attention_type=attention_type,
+            mid_block_only_cross_attention=mid_block_only_cross_attention,
+            cross_attention_norm=cross_attention_norm,
+            addition_embed_type_num_heads=addition_embed_type_num_heads,
+        )
+
+        # 2 - Do model surgery on control model
+        # 2.1 - Allow to use the same time information as the base model
+        adjust_time_dims(self.control_model, time_embedding_input_dim, time_embedding_dim)
+
+        # 2.2 - Allow for information infusion from base model
+        def compute_block_out_channels(subblock_channels, layers_per_block):
+            channels = []
+            for i, (_, subblock_out_channels) in enumerate(subblock_channels):
+                # first subblock is the conv_in
+                if i == 0:
+                    continue
+                # every block consists of `layers_per_block` resnet/attention subblocks and a down sample subblock
+                if i % (layers_per_block + 1) == 0:
+                    channels.append(subblock_out_channels)
+                # the last block doesn't have a down conv, so is handled separately
+                if i == len(subblock_channels) - 1:
+                    channels.append(subblock_out_channels)
+            return channels
+
+        base_block_out_channels = compute_block_out_channels(
+            subblock_channels=base_model_channel_sizes["down"], layers_per_block=layers_per_block
+        )
+
+        extra_channels = list(
+            zip(base_block_out_channels[0:1] + base_block_out_channels[:-1], base_block_out_channels)
+        )
+        for i, (e1, e2) in enumerate(extra_channels):
+            increase_block_input_in_encoder_resnet(self.control_model, block_no=i, resnet_idx=0, by=e1)
+            increase_block_input_in_encoder_resnet(self.control_model, block_no=i, resnet_idx=1, by=e2)
+            if self.control_model.down_blocks[i].downsamplers:
+                increase_block_input_in_encoder_downsampler(self.control_model, block_no=i, by=e2)
+        increase_block_input_in_mid_resnet(self.control_model, by=base_block_out_channels[-1])
```

Comment:
A comment would be nice.

Reply:
I've now simplified the code and commented more